### PR TITLE
aggregate enhancement

### PR DESCRIPTION
**Pull request by czpmango**
support operation between aggs.

Fixes [https://jira.nebula-graph.io/browse/NG-516](https://jira.nebula-graph.io/browse/NG-516)
Fixes [https://github.com/vesoft-inc/nebula-graph/issues/846](https://github.com/vesoft-inc/nebula-graph/issues/846)

### CODE REVIEW

File: src/validator/GroupByValidator.cpp [hunk 1]
```diff
@@ -34,38 +34,55 @@ Status GroupByValidator::validateYield(const YieldClause* yieldClause) {
...
+            }
         }
-        auto colExpr = col->expr();
+
         // collect exprs for check later
         if (colExpr->kind() == Expression::Kind::kAggregate) {
             auto* aggExpr = static_cast<AggregateExpression*>(colExpr);
             NG_RETURN_IF_ERROR(ExpressionUtils::checkAggExpr(aggExpr));
-        } else {
+        } else if (!ExpressionUtils::isEvaluableExpr(colExpr)) {
```
Comment (CPWstatic) [inline src/validator/GroupByValidator.cpp:67]: Why we have this restriction？

File: tests/tck/features/aggregate/Agg.feature [hunk 2]
```diff
@@ -384,6 +384,18 @@ Feature: Basic Aggregate and GroupBy
       | dst             | count |
       | "Tony Parker"   | 1     |
       | "Manu Ginobili" | 1     |
+    When executing query:
+      """
+      GO FROM "Tim Duncan" OVER like YIELD like._dst AS dst, $$.player.age AS age
+      | GROUP BY $-.dst
+      YIELD
+        $-.dst AS dst,
+        (INT)(sum($-.age)/count($-.age))+avg(distinct $-.age+1)+1 AS statistics
```
Comment (CPWstatic) [inline tests/tck/features/aggregate/Agg.feature:393]: Do we have some exception cases such as `sum(count())`?

Comment (czpmango) [reply tests/tck/features/aggregate/Agg.feature:393]: Yep, in Agg.feature `error check` scenario.

File: src/validator/GroupByValidator.cpp (hunk 1 above)
Comment (czpmango) [reply src/validator/GroupByValidator.cpp:67]: Avoid constant expressions as `groupKey` when doing implicit groupby.

File: src/validator/MatchValidator.cpp [hunk 3]
```diff
@@ -685,30 +685,35 @@ Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
...
+                                                 colExpr->toString().c_str());
+                }
+
+                yieldCtx.groupItems_.emplace_back(
+                    yieldCtx.qctx->objPool()->add(agg->clone().release()));
+
                 yieldCtx.needGenProject_ = true;
+                yieldCtx.aggOutputColumnNames_.emplace_back(agg->toString());
+            }
+            if (aggs.size()) {
```
Comment (Shylock-Hg) [inline src/validator/MatchValidator.cpp:708]: !aggs.empty()

Comment (czpmango) [reply src/validator/MatchValidator.cpp:708]: ACK.

**Review by yixinglu — COMMENTED** [review]
Why not consider to merge the aggregation validation for groupby and match validators?

**Comment by czpmango** [comment]
> Why not consider to merge the aggregation validation for groupby and match validators?

There are many differences between ngql and cypher, such as:
 - ngql's schema check is stricter than cypher
 - cypher retains more execution information than ngql when doing Multi-step-matching
 - cypher only support implicit-groupby, ngql's groupby is more like mysql
 - ...
 
Some of the above limitations cannot be circumvented at the implementation layer.

Although the implementation is separated at present, the common behavior has been abstracted as util-functions as possible to ensure the DRY principle, such as some expression-related processing.

However, it is still possible to do some refactoring to unify their implementation in the future, but it can not be merged now.

**Review by yixinglu — APPROVED** [review]

File: tests/tck/features/aggregate/Agg.feature [hunk 4]
```diff
@@ -418,6 +430,17 @@ Feature: Basic Aggregate and GroupBy
       | 45  | 13    |
       | 46  | 2     |
       | 47  | 3     |
+    When executing query:
```
Comment (Shylock-Hg) [inline tests/tck/features/aggregate/Agg.feature:433]: Do you have a case about nested agg call.

Comment (czpmango) [reply tests/tck/features/aggregate/Agg.feature:433]: Same as https://github.com/vesoft-inc/nebula-graph/pull/968#discussion_r616303222 ?

**Review by jievince — APPROVED** [review]

**Pull request merged**